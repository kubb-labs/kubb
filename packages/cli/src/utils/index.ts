export * from './getConfig.ts'
export * from './getPlugins.ts'
export * from './watcher.ts'
export * from './getCosmiConfig.ts'
export * from './parseHrtimeToSeconds.ts'
export * from './parseText.ts'
