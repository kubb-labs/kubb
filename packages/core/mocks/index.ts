import { pascalCase } from '../src/transformers/casing.ts'

import type { PluginManager } from '../src/PluginManager.ts'
import { readSync } from '../src/fs/read.ts'

export const mockedPluginManager = {
  resolveName: ({ name, type }) => {
    if (type === 'type') {
      return pascalCase(name)
    }

    return name
  },
  resolvePath: ({ baseName }) => baseName,
  logger: {
    emit(message) {
      console.log(message)
    },
    on(eventName, args) {},
    logLevel: 'info',
  },
  getFile: ({ name, extName, pluginKey }) => {
    const baseName = `${name}${extName}`
    let source = ''

    try {
      source = readSync(baseName)
    } catch (_e) {
      //
    }

    return {
      path: baseName,
      baseName,
      meta: {
        pluginKey,
      },
      source,
    }
  },
} as PluginManager
