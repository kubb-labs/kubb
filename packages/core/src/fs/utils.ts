import { normalize, relative } from 'node:path'

function slash(path: string, platform: 'windows' | 'mac' | 'linux' = 'linux') {
  const isWindowsPath = /^\\\\\?\\/.test(path)
  const normalizedPath = normalize(path)

  if (['linux', 'mac'].includes(platform) && !isWindowsPath) {
    // linux and mac
    return normalizedPath.replaceAll(/\\/g, '/').replace('../', '')
  }

  // windows
  return normalizedPath.replaceAll(/\\/g, '/').replace('../', '')
}

export function getRelativePath(rootDir?: string | null, filePath?: string | null, platform: 'windows' | 'mac' | 'linux' = 'linux'): string {
  if (!rootDir || !filePath) {
    throw new Error(`Root and file should be filled in when retrieving the relativePath, ${rootDir || ''} ${filePath || ''}`)
  }

  const relativePath = relative(rootDir, filePath)

  // On Windows, paths are separated with a "\"
  // However, web browsers use "/" no matter the platform
  const slashedPath = slash(relativePath, platform)

  if (slashedPath.startsWith('../')) {
    return slashedPath
  }

  return `./${slashedPath}`
}
