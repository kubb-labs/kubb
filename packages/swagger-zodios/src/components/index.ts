export { Definitions } from './Definitions.tsx'
