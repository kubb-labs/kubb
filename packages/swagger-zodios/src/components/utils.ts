import transformers from '@kubb/core/transformers'

import type { ResolveNameParams, ResolvePathParams } from '@kubb/core'
import type { KubbFile } from '@kubb/core'
import type { OperationsByMethod, OperationSchemas } from '@kubb/swagger'
import type { OasTypes, Operation } from '@kubb/swagger/oas'

const parameters = {
  getPathParams(name: string, item: NonNullable<OperationSchemas['pathParams']>): string[] {
    const parameters: string[] = []

    item.keys?.forEach((key) => {
      const schema = item.schema?.properties?.[key] as OasTypes.SchemaObject
      const shape = item.schema?.$ref ? `schema.shape['${key}']` : `shape['${key}']`
      const required = Array.isArray(item.schema.required) ? !!item.schema.required?.length : !!item.schema.required
      const definedSchema = required ? `${name}.${shape}` : `${name}.unwrap().${shape}`

      parameters.push(`
        {
          name: "${key}",
          description: \`${transformers.escape(schema?.description)}\`,
          type: "Path",
          schema: ${definedSchema}
        }
      `)
    })

    return parameters
  },
  getQueryParams(name: string, item: NonNullable<OperationSchemas['queryParams']>): string[] {
    const parameters: string[] = []

    item.keys?.forEach((key) => {
      const schema = item.schema?.properties?.[key] as OasTypes.SchemaObject
      const shape = item.schema?.$ref ? `schema.shape['${key}']` : `shape['${key}']`
      const required = Array.isArray(item.schema.required) ? !!item.schema.required?.length : !!item.schema.required
      const definedSchema = required ? `${name}.${shape}` : `${name}.unwrap().${shape}`

      parameters.push(`
      {
        name: "${key}",
        description: \`${transformers.escape(schema?.description)}\`,
        type: "Query",
        schema: ${definedSchema}
      }
    `)
    })
    return parameters
  },
  getHeaderParams(name: string, item: NonNullable<OperationSchemas['queryParams']>): string[] {
    const parameters: string[] = []

    item.keys?.forEach((key) => {
      const schema = item.schema?.properties?.[key] as OasTypes.SchemaObject
      const shape = item.schema?.$ref ? `schema.shape['${key}']` : `shape['${key}']`
      const required = Array.isArray(item.schema.required) ? !!item.schema.required?.length : !!item.schema.required
      const definedSchema = required ? `${name}.${shape}` : `${name}.unwrap().${shape}`

      parameters.push(`
      {
        name: "${key}",
        description: \`${transformers.escape(schema?.description)}\`,
        type: "Header",
        schema: ${definedSchema}
      }
    `)
    })
    return parameters
  },
  getRequest(name: string, item: NonNullable<OperationSchemas['request']>): string[] {
    const parameters: string[] = []

    parameters.push(`
    {
      name: "${item.name}",
      description: \`${transformers.escape(item.description)}\`,
      type: "Body",
      schema: ${name}
    }
  `)
    return parameters
  },
} as const

export function getDefinitionsImports(
  operationsByMethod: OperationsByMethod,
  {
    resolveName,
    resolvePath,
    pluginKey,
  }: {
    resolveName: (params: ResolveNameParams) => string
    resolvePath: (params: ResolvePathParams) => KubbFile.OptionalPath
    pluginKey: ResolveNameParams['pluginKey']
  },
): Array<{ name: string; path: KubbFile.OptionalPath }> {
  const definitions: Array<{ name: string; operation: Operation }> = []

  Object.keys(operationsByMethod).forEach((path) => {
    const operations = operationsByMethod[path]
    const filteredOperations = [operations?.get, operations?.post, operations?.patch, operations?.put, operations?.delete].filter(Boolean)

    filteredOperations.forEach(({ operation, schemas }) => {
      const responseName = resolveName({
        name: schemas.response.name,
        pluginKey,
        type: 'function',
      })

      definitions.push({ name: responseName, operation })

      if (schemas.pathParams?.name) {
        const name = resolveName({
          name: schemas.pathParams.name,
          pluginKey,
          type: 'function',
        })

        definitions.push({ name, operation })
      }

      if (schemas.queryParams?.name) {
        const name = resolveName({
          name: schemas.queryParams.name,
          pluginKey,
          type: 'function',
        })

        definitions.push({ name, operation })
      }

      if (schemas.headerParams?.name) {
        const name = resolveName({
          name: schemas.headerParams.name,
          pluginKey,
          type: 'function',
        })

        definitions.push({ name, operation })
      }

      if (schemas.request?.name) {
        const name = resolveName({
          name: schemas.request.name,
          pluginKey,
          type: 'function',
        })

        definitions.push({ name, operation })
      }
      if (schemas.errors) {
        schemas.errors.forEach((errorOperationSchema) => {
          if (!errorOperationSchema.statusCode) {
            return
          }

          const name = resolveName({
            name: `${operation.getOperationId()} ${errorOperationSchema.statusCode}`,
            pluginKey,
            type: 'function',
          })

          definitions.push({ name, operation })
        })
      }
    })
  })

  return definitions.map(({ name, operation }) => {
    const baseName = resolveName({
      name: `${operation.getOperationId()}`,
      pluginKey,
      type: 'file',
    })

    const path = resolvePath({
      pluginKey,
      baseName: `${baseName}.ts`,
      options: {
        tag: operation?.getTags()[0]?.name,
      },
    })
    return { name, path }
  })
}

export function getDefinitions(
  operationsByMethod: OperationsByMethod,
  {
    resolveName,
    pluginKey,
  }: {
    resolveName: (params: ResolveNameParams) => string
    pluginKey: ResolveNameParams['pluginKey']
  },
): Array<{
  operation: Operation
  response: string | undefined
  parameters: string[]
  errors: string[]
}> {
  const definitions: Array<{
    response: string
    operation: Operation
    parameters: string[]
    errors: string[]
  }> = []

  Object.keys(operationsByMethod).forEach((path) => {
    const operations = operationsByMethod[path]
    const filteredOperations = [operations?.get, operations?.post, operations?.patch, operations?.put, operations?.delete].filter(Boolean)

    filteredOperations.forEach(({ operation, schemas }) => {
      let params: string[] = []
      const errors: string[] = []
      const responseName = resolveName({
        name: schemas.response.name,
        pluginKey,
        type: 'function',
      })

      if (schemas.pathParams?.name) {
        const name = resolveName({
          name: schemas.pathParams.name,
          pluginKey,
          type: 'function',
        })
        params = [...params, ...parameters.getPathParams(name, schemas.pathParams)]
      }

      if (schemas.queryParams?.name) {
        const name = resolveName({
          name: schemas.queryParams.name,
          pluginKey,
          type: 'function',
        })
        params = [...params, ...parameters.getQueryParams(name, schemas.queryParams)]
      }

      if (schemas.headerParams?.name) {
        const name = resolveName({
          name: schemas.headerParams.name,
          pluginKey,
          type: 'function',
        })
        params = [...params, ...parameters.getHeaderParams(name, schemas.headerParams)]
      }

      if (schemas.request?.name) {
        const name = resolveName({
          name: schemas.request.name,
          pluginKey,
          type: 'function',
        })
        params = [...params, ...parameters.getRequest(name, schemas.request)]
      }
      if (schemas.errors) {
        schemas.errors.forEach((errorOperationSchema) => {
          if (!errorOperationSchema.statusCode) {
            return
          }

          const name = resolveName({
            name: `${operation.getOperationId()} ${errorOperationSchema.statusCode}`,
            pluginKey,
            type: 'function',
          })

          if (errorOperationSchema.statusCode) {
            errors.push(`
              {
                status: ${errorOperationSchema.statusCode},
                description: \`${transformers.escape(errorOperationSchema.description)}\`,
                schema: ${name}
              }
            `)
          }
        })
      }
      definitions.push({
        operation,
        parameters: params,
        errors,
        response: responseName,
      })
    })
  })

  return definitions
}
