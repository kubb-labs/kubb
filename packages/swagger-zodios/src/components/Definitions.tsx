import transformers from '@kubb/core/transformers'
import { URLPath } from '@kubb/core/utils'
import { Editor, File, useGetFile, usePlugin } from '@kubb/react'
import { usePluginManager } from '@kubb/react'
import { pluginKey as swaggerZodPluginKey } from '@kubb/swagger-zod'

import { getDefinitions, getDefinitionsImports } from './utils.ts'

import type { OperationsByMethod } from '@kubb/swagger'
import type { ReactNode } from 'react'
import type { FileMeta, PluginOptions } from '../types.ts'

type TemplateProps = {
  /**
   * Name of the function
   */
  name: string
  definitions: string[]
  baseURL: string | undefined
}

function Template({ name, definitions, baseURL }: TemplateProps): ReactNode {
  return (
    <>
      {`export const endpoints = makeApi([${definitions.join(',')}])`}
      <br />
      {'export const getAPI = (baseUrl: string) => new Zodios(baseUrl, endpoints)'}
      <br />
      {baseURL && `export const ${name} = new Zodios('${baseURL}', endpoints)`}
      {!baseURL && `export const  ${name} = new Zodios(endpoints)`}
      <br />
      {`export default ${name}`}
    </>
  )
}
const defaultTemplates = { default: Template } as const

type Props = {
  baseURL: string | undefined
  /**
   * @deprecated
   */
  operationsByMethod: OperationsByMethod
  /**
   * This will make it possible to override the default behaviour.
   */
  Template?: React.ComponentType<React.ComponentProps<typeof Template>>
}

export function Definitions({ baseURL, operationsByMethod, Template = defaultTemplates.default }: Props): ReactNode {
  const pluginManager = usePluginManager()
  const definitions = getDefinitions(operationsByMethod, {
    resolveName: pluginManager.resolveName,
    pluginKey: swaggerZodPluginKey,
  })

  return (
    <Template
      name={'api'}
      baseURL={baseURL}
      definitions={definitions.map(({ errors, response, operation, parameters }) => {
        return `
        {
          method: "${operation.method}",
          path: "${new URLPath(operation.path).URL}",
          description: \`${transformers.escape(operation.getDescription())}\`,
          requestFormat: "json",
          parameters: [
              ${parameters.join(',')}
          ],
          response: ${response},
          errors: [
              ${errors.join(',')}
          ],
      }
      `
      })}
    />
  )
}

type FileProps = {
  name: string
  baseURL: string | undefined
  /**
   * @deprecated
   */
  operationsByMethod: OperationsByMethod
  /**
   * This will make it possible to override the default behaviour.
   */
  templates?: typeof defaultTemplates
}

Definitions.File = function ({ name, baseURL, operationsByMethod, templates = defaultTemplates }: FileProps): ReactNode {
  const pluginManager = usePluginManager()
  const { key: pluginKey } = usePlugin<PluginOptions>()
  const file = useGetFile({ name, extName: '.ts', pluginKey })

  const definitionsImports = getDefinitionsImports(operationsByMethod, {
    resolveName: pluginManager.resolveName,
    resolvePath: pluginManager.resolvePath,
    pluginKey: swaggerZodPluginKey,
  })

  const imports = definitionsImports
    .map(({ name, path }, index) => {
      if (!path) {
        return null
      }

      return <File.Import key={index} name={[name]} root={file.path} path={path} />
    })
    .filter(Boolean)

  const Template = templates.default

  return (
    <Editor language="typescript">
      <File<FileMeta> baseName={file.baseName} path={file.path} meta={file.meta}>
        <File.Import name={['makeApi', 'Zodios']} path="@zodios/core" />
        {imports}
        <File.Source>
          <Definitions Template={Template} operationsByMethod={operationsByMethod} baseURL={baseURL} />
        </File.Source>
      </File>
    </Editor>
  )
}

Definitions.templates = defaultTemplates
