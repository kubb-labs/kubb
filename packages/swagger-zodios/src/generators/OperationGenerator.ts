import { getRelativePath, transformers, URLPath } from '@kubb/core/utils'
import { OperationGenerator as Generator, resolve } from '@kubb/swagger'
import { pluginKey as swaggerZodPluginKey } from '@kubb/swagger-zod'

import { camelCase, camelCaseTransformMerge } from 'change-case'

import type { KubbFile } from '@kubb/core'
import type { HttpMethod, OpenAPIV3, Operation, Resolver } from '@kubb/swagger'

type Options = {
  output: string
}

const methods: HttpMethod[] = ['get', 'post', 'patch', 'put', 'delete']

export class OperationGenerator extends Generator<Options> {
  resolve(): Resolver {
    const { output } = this.options
    const { pluginManager, plugin } = this.context

    const name = pluginManager.resolveName({ name: output.replace('.ts', ''), pluginKey: plugin.key })

    return resolve({
      name,
      resolveName: pluginManager.resolveName,
      resolvePath: pluginManager.resolvePath,
      pluginKey: plugin.key,
    })
  }

  resolveResponse(operation: Operation): Resolver {
    const { pluginManager } = this.context

    const schemas = this.getSchemas(operation)

    const name = pluginManager.resolveName({ name: schemas.response.name, pluginKey: swaggerZodPluginKey })
    const baseName: KubbFile.BaseName = `${camelCase(`${operation.getOperationId()}Schema`, { delimiter: '', transform: camelCaseTransformMerge })}.ts`
    const filePath = pluginManager.resolvePath({
      baseName: baseName,
      options: { tag: operation.getTags()[0]?.name },
      pluginKey: swaggerZodPluginKey,
    })

    if (!filePath || !name) {
      throw new Error('Filepath should be defined')
    }

    return {
      name,
      baseName: baseName,
      path: filePath,
    }
  }

  resolveRequest(operation: Operation): Resolver {
    const { pluginManager } = this.context

    const schemas = this.getSchemas(operation)

    if (!schemas.request?.name) {
      throw new Error('schemas.request should be defined')
    }

    const name = pluginManager.resolveName({ name: schemas.request.name, pluginKey: swaggerZodPluginKey })
    const baseName = `${camelCase(`${operation.getOperationId()}Schema`, { delimiter: '', transform: camelCaseTransformMerge })}.ts` as const
    const filePath = pluginManager.resolvePath({
      baseName: baseName,
      options: { tag: operation.getTags()[0]?.name },
      pluginKey: swaggerZodPluginKey,
    })

    if (!filePath || !name) {
      throw new Error('Filepath should be defined')
    }

    return {
      name,
      baseName: baseName,
      path: filePath,
    }
  }

  resolveHeaderParams(operation: Operation): Resolver {
    const { pluginManager } = this.context

    const schemas = this.getSchemas(operation)
    if (!schemas.headerParams?.name) {
      throw new Error('schemas.pathParams should be defined')
    }

    const name = pluginManager.resolveName({ name: schemas.headerParams.name, pluginKey: swaggerZodPluginKey })
    const baseName = `${camelCase(`${operation.getOperationId()}Schema`, { delimiter: '', transform: camelCaseTransformMerge })}.ts` as const
    const filePath = pluginManager.resolvePath({
      baseName: baseName,
      options: { tag: operation.getTags()[0]?.name },
      pluginKey: swaggerZodPluginKey,
    })

    if (!filePath || !name) {
      throw new Error('Filepath should be defined')
    }

    return {
      name,
      baseName: baseName,
      path: filePath,
    }
  }

  resolvePathParams(operation: Operation): Resolver {
    const { pluginManager } = this.context

    const schemas = this.getSchemas(operation)
    if (!schemas.pathParams?.name) {
      throw new Error('schemas.pathParams should be defined')
    }

    const name = pluginManager.resolveName({ name: schemas.pathParams.name, pluginKey: swaggerZodPluginKey })
    const baseName = `${camelCase(`${operation.getOperationId()}Schema`, { delimiter: '', transform: camelCaseTransformMerge })}.ts` as const
    const filePath = pluginManager.resolvePath({
      baseName: baseName,
      options: { tag: operation.getTags()[0]?.name },
      pluginKey: swaggerZodPluginKey,
    })

    if (!filePath || !name) {
      throw new Error('Filepath should be defined')
    }

    return {
      name,
      baseName: baseName,
      path: filePath,
    }
  }

  resolveQueryParams(operation: Operation): Resolver {
    const { pluginManager } = this.context

    const schemas = this.getSchemas(operation)

    if (!schemas.queryParams?.name) {
      throw new Error('schemas.queryParams should be defined')
    }

    const name = pluginManager.resolveName({ name: schemas.queryParams.name, pluginKey: swaggerZodPluginKey })
    const baseName = `${camelCase(`${operation.getOperationId()}Schema`, { delimiter: '', transform: camelCaseTransformMerge })}.ts` as const
    const filePath = pluginManager.resolvePath({
      baseName: baseName,
      options: { tag: operation.getTags()[0]?.name },
      pluginKey: swaggerZodPluginKey,
    })

    if (!filePath || !name) {
      throw new Error('Filepath should be defined')
    }

    return {
      name,
      baseName: baseName,
      path: filePath,
    }
  }

  resolveError(operation: Operation, statusCode: number): Resolver {
    const { pluginManager } = this.context

    const name = pluginManager.resolveName({ name: `${operation.getOperationId()} ${statusCode}`, pluginKey: swaggerZodPluginKey })
    const baseName = `${camelCase(`${operation.getOperationId()}Schema`, { delimiter: '', transform: camelCaseTransformMerge })}.ts` as const
    const filePath = pluginManager.resolvePath({
      baseName: baseName,
      options: { tag: operation.getTags()[0]?.name },
      pluginKey: swaggerZodPluginKey,
    })

    if (!filePath || !name) {
      throw new Error('Filepath should be defined')
    }

    return {
      name,
      baseName: baseName,
      path: filePath,
    }
  }

  async all(paths: Record<string, Record<HttpMethod, Operation | undefined>>): Promise<KubbFile.File | null> {
    const imports: Array<KubbFile.Import> = [
      {
        name: ['makeApi', 'Zodios'],
        path: '@zodios/core',
      },
    ]

    const zodios = this.resolve()

    const mapOperationToZodios = (operation: Operation): string => {
      const schemas = this.getSchemas(operation)
      const parameters: string[] = []
      const errors: string[] = []

      const response = this.resolveResponse(operation)

      imports.push({
        name: [response.name],
        path: getRelativePath(zodios.path, response.path),
      })

      if (schemas.pathParams) {
        const pathParams = this.resolvePathParams(operation)

        imports.push({
          name: [pathParams.name],
          path: getRelativePath(zodios.path, pathParams.path),
        })

        schemas.pathParams.keys?.forEach((key) => {
          const schema = schemas.pathParams?.schema?.properties?.[key] as OpenAPIV3.SchemaObject
          const zodSchema = schemas.pathParams?.schema?.$ref ? `${pathParams.name}.schema.shape['${key}']` : `${pathParams.name}.shape['${key}']`

          parameters.push(`
          {
            name: "${key}",
            description: \`${transformers.escape(schema?.description)}\`,
            type: "Path",
            schema: ${zodSchema}
          }
        `)
        })
      }

      if (schemas.queryParams) {
        const queryParams = this.resolveQueryParams(operation)

        imports.push({
          name: [queryParams.name],
          path: getRelativePath(zodios.path, queryParams.path),
        })

        schemas.queryParams.keys?.forEach((key) => {
          const schema = schemas.queryParams?.schema?.properties?.[key] as OpenAPIV3.SchemaObject
          const zodSchema = schemas.queryParams?.schema?.$ref ? `${queryParams.name}.schema.shape['${key}']` : `${queryParams.name}.shape['${key}']`

          parameters.push(`
          {
            name: "${key}",
            description: \`${transformers.escape(schema?.description)}\`,
            type: "Query",
            schema: ${zodSchema}
          }
        `)
        })
      }

      if (schemas.request) {
        const requestBody = this.resolveRequest(operation)

        imports.push({
          name: [requestBody.name],
          path: getRelativePath(zodios.path, requestBody.path),
        })

        parameters.push(`
        {
          name: "${schemas.request.name}",
          description: \`${transformers.escape(schemas.request.description)}\`,
          type: "Body",
          schema: ${requestBody.name}
        }
        `)
      }

      if (schemas.headerParams) {
        const headerParams = this.resolveHeaderParams(operation)

        imports.push({
          name: [headerParams.name],
          path: getRelativePath(zodios.path, headerParams.path),
        })

        schemas.headerParams.keys?.forEach((key) => {
          const schema = schemas.headerParams?.schema?.properties?.[key] as OpenAPIV3.SchemaObject
          const zodSchema = schemas.headerParams?.schema?.$ref ? `${headerParams.name}.schema.shape['${key}']` : `${headerParams.name}.shape['${key}']`

          parameters.push(`
          {
            name: "${key}",
            description: \`${transformers.escape(schema?.description)}\`,
            type: "Header",
            schema: ${zodSchema}
          }
        `)
        })
      }

      if (schemas.errors) {
        schemas.errors.forEach((errorOperationSchema) => {
          if (!errorOperationSchema.statusCode) {
            return
          }

          const { path: filePath, name } = this.resolveError(operation, errorOperationSchema.statusCode)

          imports.push({
            name: [name],
            path: getRelativePath(zodios.path, filePath),
          })

          if (errorOperationSchema.statusCode) {
            errors.push(`
              {
                status: ${errorOperationSchema.statusCode},
                description: \`${transformers.escape(errorOperationSchema.description)}\`,
                schema: ${name}
              }
            `)
          }
        })
      }

      return `
        {
          method: "${operation.method}",
          path: "${new URLPath(operation.path).URL}",
          description: \`${transformers.escape(operation.getDescription())}\`,
          requestFormat: "json",
          parameters: [
              ${parameters.join(',')}
          ],
          response: ${response.name},
          errors: [
              ${errors.join(',')}
          ],
      }
      `
    }
    const definitions = Object.keys(paths).reduce((acc, path) => {
      const operations = paths[path]

      if (operations) {
        methods.forEach((method) => {
          // use isSkipped to also exclude operations(skipby in our Zod plugin).
          if (operations[method] && !this.isSkipped(operations[method]!, method)) {
            acc.push(mapOperationToZodios(operations[method]!))
          }
        })
      }

      return acc
    }, [] as string[])

    const sources: string[] = []

    sources.push(`
      const endpoints = makeApi([${definitions.join(',')}]);

      export const api = new Zodios(endpoints);

      export default api;
    `)

    return {
      path: zodios.path,
      baseName: zodios.baseName,
      source: sources.join('\n'),
      imports,
    }
  }

  async get(): Promise<KubbFile.File | null> {
    return null
  }

  async post(): Promise<KubbFile.File | null> {
    return null
  }
  async patch(): Promise<KubbFile.File | null> {
    return null
  }

  async put(): Promise<KubbFile.File | null> {
    return null
  }

  async delete(): Promise<KubbFile.File | null> {
    return null
  }
}
