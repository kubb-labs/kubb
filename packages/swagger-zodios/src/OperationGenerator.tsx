import { createRoot } from '@kubb/react'
import { OperationGenerator as Generator } from '@kubb/swagger'
import { Oas } from '@kubb/swagger/components'

import { Definitions } from './components/Definitions.tsx'

import type { AppContextProps } from '@kubb/react'
import type { OperationMethodResult, OperationsByMethod } from '@kubb/swagger'
import type { Operation } from '@kubb/swagger/oas'
import type { FileMeta, PluginOptions } from './types.ts'

export class OperationGenerator extends Generator<PluginOptions['resolvedOptions'], PluginOptions> {
  async all(operations: Operation[], operationsByMethod: OperationsByMethod): OperationMethodResult<FileMeta> {
    const { oas, pluginManager, plugin } = this.context

    const root = createRoot<AppContextProps<PluginOptions['appMeta']>>({
      logger: pluginManager.logger,
    })

    root.render(
      <Oas oas={oas} operations={operations} getOperationSchemas={(...props) => this.getSchemas(...props)}>
        <Definitions.File name={this.options.name} baseURL={this.options.baseURL} operationsByMethod={operationsByMethod} />
      </Oas>,
      {
        meta: { pluginManager, plugin },
      },
    )

    return root.files
  }

  async operation(): OperationMethodResult<FileMeta> {
    return null
  }

  async get(): OperationMethodResult<FileMeta> {
    return null
  }

  async post(): OperationMethodResult<FileMeta> {
    return null
  }
  async patch(): OperationMethodResult<FileMeta> {
    return null
  }

  async put(): OperationMethodResult<FileMeta> {
    return null
  }

  async delete(): OperationMethodResult<FileMeta> {
    return null
  }
}
