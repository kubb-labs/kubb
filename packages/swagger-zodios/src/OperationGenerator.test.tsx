import { mockedPluginManager } from '@kubb/core/mocks'

import { OperationGenerator } from './OperationGenerator.tsx'

import type { Plugin } from '@kubb/core'
import type { GetOperationGeneratorOptions } from '@kubb/plugin-oas'
import { parseFromConfig } from '@kubb/plugin-oas/utils'
import type { PluginZodios } from './types.ts'
import { Oas } from '@kubb/plugin-oas/components'
import { App } from '@kubb/react'
import { createRootServer } from '@kubb/react/server'
import { Definitions } from './components/Definitions.tsx'

describe('OperationGenerator', async () => {
  const oas = await parseFromConfig({
    root: './',
    output: {
      path: 'test',
      clean: true,
    },
    input: { path: 'packages/swagger-zodios/mocks/petStore.yaml' },
  })

  test('includes an alias property when includeOperationIdAsAlias is true', async () => {
    const options: GetOperationGeneratorOptions<OperationGenerator> = {
      baseURL: '',
      includeOperationIdAsAlias: true,
      name: 'example',
      extName: undefined
    }
    const plugin = { options } as Plugin<PluginZodios>

    const generator = new OperationGenerator(options, {
      oas,
      exclude: [],
      include: [{ type: 'operationId', pattern: 'listPets' }],
      pluginManager: mockedPluginManager,
      plugin,
      contentType: undefined,
      override: undefined,
      mode: 'split',
    })

    // Need to build to set the operationsByMethod
    generator.build()

    const operation = oas.operation('/pets', 'get')

    const Component = () => {
      return (
        <App plugin={plugin} pluginManager={mockedPluginManager} mode="split">
          <Oas oas={oas} operations={[operation]} generator={generator}>
            <Oas.Operation operation={operation}>
              <Definitions.File operationsByMethod={generator.operationsByMethod} {...options} />
            </Oas.Operation>
          </Oas>
        </App>
      )
    }
    const root = createRootServer({ logger: mockedPluginManager.logger })
    const output = await root.renderToString(<Component />)

    expect(output).toMatchSnapshot()
  })

  test('does not include an alias property when includeOperationIdAsAlias is false', async () => {
    const options: GetOperationGeneratorOptions<OperationGenerator> = {
      baseURL: '',
      includeOperationIdAsAlias: false,
      name: 'example',
      extName: undefined
    }
    const plugin = { options } as Plugin<PluginZodios>

    const generator = new OperationGenerator(options, {
      oas,
      exclude: [],
      include: [{ type: 'operationId', pattern: 'listPets' }],
      pluginManager: mockedPluginManager,
      plugin,
      contentType: undefined,
      override: undefined,
      mode: 'split',
    })

    // Need to build to set the operationsByMethod
    generator.build()

    const operation = oas.operation('/pets', 'get')

    const Component = () => {
      return (
        <App plugin={plugin} pluginManager={mockedPluginManager} mode="split">
          <Oas oas={oas} operations={[operation]} generator={generator}>
            <Oas.Operation operation={operation}>
              <Definitions.File operationsByMethod={generator.operationsByMethod} {...options} />
            </Oas.Operation>
          </Oas>
        </App>
      )
    }
    const root = createRootServer({ logger: mockedPluginManager.logger })
    const output = await root.renderToString(<Component />)

    expect(output).toMatchSnapshot()
  })
})
