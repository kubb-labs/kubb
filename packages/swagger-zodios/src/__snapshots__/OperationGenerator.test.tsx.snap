// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OperationGenerator > does not include an alias property when includeOperationIdAsAlias is false 1`] = `
"export const endpoints = makeApi([
  {
    method: 'get',
    path: '/pets',
    description: \`\`,
    requestFormat: 'json',
    parameters: [
      {
        name: 'limit',
        description: \`How many items to return at one time (max 100)\`,
        type: 'Query',
        schema: ListPetsQueryParams.unwrap().shape['limit'],
      },
    ],
    response: ListPetsQueryResponse,
    errors: [],
  },
])
export const getAPI = (baseUrl: string) => new Zodios(baseUrl, endpoints)
export const api = new Zodios(endpoints)
export default api
"
`;

exports[`OperationGenerator > includes an alias property when includeOperationIdAsAlias is true 1`] = `
"export const endpoints = makeApi([
  {
    method: 'get',
    path: '/pets',
    alias: 'listPets',
    description: \`\`,
    requestFormat: 'json',
    parameters: [
      {
        name: 'limit',
        description: \`How many items to return at one time (max 100)\`,
        type: 'Query',
        schema: ListPetsQueryParams.unwrap().shape['limit'],
      },
    ],
    response: ListPetsQueryResponse,
    errors: [],
  },
])
export const getAPI = (baseUrl: string) => new Zodios(baseUrl, endpoints)
export const api = new Zodios(endpoints)
export default api
"
`;
